File: Pod/Classes/ios/NYTPhotosOverlayView.m
//
//  NYTPhotosOverlayView.m
//  NYTPhotoViewer
//
//  Created by Brian Capps on 2/17/15.
//
//

#import "NYTPhotosOverlayView.h"
#import "NYTPhotoCaptionViewLayoutWidthHinting.h"
#import "NYTLoadingIndicatorView.h"

@interface NYTPhotosOverlayView ()

@property (nonatomic) UINavigationItem *navigationItem;
@property (nonatomic) UINavigationBar *navigationBar;
@property (nonatomic) CAGradientLayer *gradientLayer;

@end

@implementation NYTPhotosOverlayView

#pragma mark - UIView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    
    if (self) {
        [self setupNavigationBar];
        [self setupLoadingIndicator];
    }
        
    return self;
}

// Pass the touches down to other views: http://stackoverflow.com/a/8104378
- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event {
    UIView *hitView = [super hitTest:point withEvent:event];
    
    if (hitView == self) {
        return nil;
    }
    
    return hitView;
}

- (void)layoutSubviews {
    // The navigation bar has a different intrinsic content size upon rotation, so we must update to that new size.
    // Do it without animation to more closely match the behavior in `UINavigationController`
    [UIView performWithoutAnimation:^{
        [self.navigationBar invalidateIntrinsicContentSize];
        [self.navigationBar layoutIfNeeded];
    }];
    
    [super layoutSubviews];
    [self layoutNavbarGradient];

    self.loadingIndicatorView.frame = (CGRect){0, 0, self.bounds.size.width, 2};
    if ([self.captionView conformsToProtocol:@protocol(NYTPhotoCaptionViewLayoutWidthHinting)]) {
        [(id<NYTPhotoCaptionViewLayoutWidthHinting>) self.captionView setPreferredMaxLayoutWidth:self.bounds.size.width];
    }
}

#pragma mark - NYTPhotosOverlayView
- (void)setupLoadingIndicator {
    self.loadingIndicatorView = [[NYTLoadingIndicatorView alloc] init];
    [self addSubview:self.loadingIndicatorView];
}

- (void)setupNavigationBar {
    self.navigationBar = [[UINavigationBar alloc] init];
    self.navigationBar.translatesAutoresizingMaskIntoConstraints = NO;
    
    // Make navigation bar background fully transparent.
    self.navigationBar.backgroundColor = [UIColor clearColor];
    self.navigationBar.barTintColor = nil;
    self.navigationBar.translucent = YES;
    self.navigationBar.shadowImage = [[UIImage alloc] init];
    [self.navigationBar setBackgroundImage:[[UIImage alloc] init] forBarMetrics:UIBarMetricsDefault];
    
    self.navigationItem = [[UINavigationItem alloc] initWithTitle:@""];
    self.navigationBar.items = @[self.navigationItem];
    
    [self addSubview:self.navigationBar];
    
    NSLayoutConstraint *topConstraint = [NSLayoutConstraint constraintWithItem:self.navigationBar attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1.0 constant:0.0];
    NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:self.navigationBar attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeWidth multiplier:1.0 constant:0.0];
    NSLayoutConstraint *horizontalPositionConstraint = [NSLayoutConstraint constraintWithItem:self.navigationBar attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0.0];
    [self addConstraints:@[topConstraint, widthConstraint, horizontalPositionConstraint]];
}

- (void)setupGradient {
    self.gradientLayer = [CAGradientLayer layer];
    self.gradientLayer.frame = self.navigationBar.layer.bounds;
    self.gradientLayer.colors = [NSArray arrayWithObjects:(id)[[UIColor blackColor] colorWithAlphaComponent:0.85].CGColor, (id)[UIColor clearColor].CGColor, nil];
    [self.navigationBar.layer insertSublayer:self.gradientLayer atIndex:1];
}

- (void)layoutNavbarGradient {
    if (self.gradientLayer) {
        self.gradientLayer.frame = self.navigationBar.bounds;
    }
}

- (void)setCaptionView:(UIView *)captionView {
    if (self.captionView == captionView) {
        return;
    }
    
    [self.captionView removeFromSuperview];
    
    _captionView = captionView;
    
    self.captionView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.captionView];
    
    NSLayoutConstraint *bottomConstraint = [NSLayoutConstraint constraintWithItem:self.captionView attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeBottom multiplier:1.0 constant:0.0];
    NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:self.captionView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeWidth multiplier:1.0 constant:0.0];
    NSLayoutConstraint *horizontalPositionConstraint = [NSLayoutConstraint constraintWithItem:self.captionView attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0.0];
    [self addConstraints:@[bottomConstraint, widthConstraint, horizontalPositionConstraint]];
}

- (UIBarButtonItem *)leftBarButtonItem {
    return self.navigationItem.leftBarButtonItem;
}

- (void)setLeftBarButtonItem:(UIBarButtonItem *)leftBarButtonItem {
    [self.navigationItem setLeftBarButtonItem:leftBarButtonItem animated:NO];
}

- (NSArray *)leftBarButtonItems {
    return self.navigationItem.leftBarButtonItems;
}

- (void)setLeftBarButtonItems:(NSArray *)leftBarButtonItems {
    [self.navigationItem setLeftBarButtonItems:leftBarButtonItems animated:NO];
}

- (UIBarButtonItem *)rightBarButtonItem {
    return self.navigationItem.rightBarButtonItem;
}

- (void)setRightBarButtonItem:(UIBarButtonItem *)rightBarButtonItem {
    [self.navigationItem setRightBarButtonItem:rightBarButtonItem animated:NO];
}

- (NSArray *)rightBarButtonItems {
    return self.navigationItem.rightBarButtonItems;
}

- (void)setRightBarButtonItems:(NSArray *)rightBarButtonItems {
    [self.navigationItem setRightBarButtonItems:rightBarButtonItems animated:NO];
}

- (NSString *)title {
    return self.navigationItem.title;
}

- (void)setTitle:(NSString *)title {
    self.navigationItem.title = title;
}

- (NSDictionary *)titleTextAttributes {
    return self.navigationBar.titleTextAttributes;
}

- (void)setTitleTextAttributes:(NSDictionary *)titleTextAttributes {
    self.navigationBar.titleTextAttributes = titleTextAttributes;
}

- (void)setNavigationBarGradient:(BOOL)navigationBarGradient {
    if (_navigationBarGradient) {
        [self.gradientLayer removeFromSuperlayer];
    }
    if (navigationBarGradient) {
        [self setupGradient];
    }
    _navigationBarGradient = navigationBarGradient;
}

@end

File: Sources/NYTPhotoViewer/NYTPhotosOverlayView.swift
//
//  NYTPhotosOverlayView.swift
//  NYTPhotoViewer
//
//  Created by Brian Capps on 2/17/15.
//  Copyright (c) 2015 NYTimes. All rights reserved.
//

import UIKit

/// A view that overlays an `NYTPhotosViewController`, and houses the left and right bar button items, a title, and a caption view.
@MainActor
public class NYTPhotosOverlayView: UIView {
    
    // MARK: - Properties
    
    /// The internal navigation bar used to set the bar button items and title of the overlay.
    public let navigationBar: UINavigationBar
    
    private let navigationItem: UINavigationItem
    
    /// The title of the overlay. Centered between the left and right bar button items.
    public var title: String? {
        get { navigationItem.title }
        set { navigationItem.title = newValue }
    }
    
    /// The attributes of the overlay's title.
    public var titleTextAttributes: [NSAttributedString.Key: Any]? {
        get { navigationBar.titleTextAttributes }
        set { navigationBar.titleTextAttributes = newValue }
    }
    
    /// The bar button item appearing at the top left of the overlay.
    public var leftBarButtonItem: UIBarButtonItem? {
        get { navigationItem.leftBarButtonItem }
        set { navigationItem.setLeftBarButton(newValue, animated: false) }
    }
    
    /// The bar button items appearing at the top left of the overlay.
    public var leftBarButtonItems: [UIBarButtonItem]? {
        get { navigationItem.leftBarButtonItems }
        set { navigationItem.setLeftBarButtonItems(newValue, animated: false) }
    }
    
    /// The bar button item appearing at the top right of the overlay.
    public var rightBarButtonItem: UIBarButtonItem? {
        get { navigationItem.rightBarButtonItem }
        set { navigationItem.setRightBarButton(newValue, animated: false) }
    }
    
    /// The bar button items appearing at the top right of the overlay.
    public var rightBarButtonItems: [UIBarButtonItem]? {
        get { navigationItem.rightBarButtonItems }
        set { navigationItem.setRightBarButtonItems(newValue, animated: false) }
    }
    
    /// A view representing the caption for the photo, which will be set to full width and locked to the bottom. Can be any `UIView` object, but is expected to respond to `intrinsicContentSize` appropriately to calculate height.
    public var captionView: UIView? {
        didSet {
            if captionView === oldValue { return }
            
            oldValue?.removeFromSuperview()
            
            guard let captionView = captionView else { return }
            
            captionView.translatesAutoresizingMaskIntoConstraints = false
            addSubview(captionView)
            
            if captionViewRespectsSafeArea {
                NSLayoutConstraint.activate([
                    captionView.bottomAnchor.constraint(equalTo: safeAreaLayoutGuide.bottomAnchor),
                    captionView.leftAnchor.constraint(equalTo: safeAreaLayoutGuide.leftAnchor),
                    captionView.rightAnchor.constraint(equalTo: safeAreaLayoutGuide.rightAnchor)
                ])
            } else {
                NSLayoutConstraint.activate([
                    captionView.bottomAnchor.constraint(equalTo: bottomAnchor),
                    captionView.widthAnchor.constraint(equalTo: widthAnchor),
                    captionView.centerXAnchor.constraint(equalTo: centerXAnchor)
                ])
            }
        }
    }
    
    /// Whether the `captionView` should respect the safe area or not
    public var captionViewRespectsSafeArea: Bool = true
    
    // MARK: - Initialization
    
    public override init(frame: CGRect) {
        self.navigationBar = UINavigationBar()
        self.navigationItem = UINavigationItem(title: "")
        
        super.init(frame: frame)
        
        setupNavigationBar()
    }
    
    required init?(coder: NSCoder) {
        self.navigationBar = UINavigationBar()
        self.navigationItem = UINavigationItem(title: "")
        
        super.init(coder: coder)
        
        setupNavigationBar()
    }
    
    // MARK: - Setup
    
    private func setupNavigationBar() {
        navigationBar.translatesAutoresizingMaskIntoConstraints = false
        
        // Make navigation bar background fully transparent.
        navigationBar.backgroundColor = .clear
        navigationBar.barTintColor = nil
        navigationBar.isTranslucent = true
        navigationBar.shadowImage = UIImage()
        navigationBar.setBackgroundImage(UIImage(), for: .default)
        
        navigationBar.items = [navigationItem]
        
        addSubview(navigationBar)
        
        NSLayoutConstraint.activate([
            navigationBar.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor),
            navigationBar.leftAnchor.constraint(equalTo: safeAreaLayoutGuide.leftAnchor),
            navigationBar.rightAnchor.constraint(equalTo: safeAreaLayoutGuide.rightAnchor)
        ])
    }
    
    // MARK: - UIView Overrides
    
    // Pass the touches down to other views
    public override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        let hitView = super.hitTest(point, with: event)
        
        if hitView == self {
            return nil
        }
        
        return hitView
    }
    
    public override func layoutSubviews() {
        // The navigation bar has a different intrinsic content size upon rotation, so we must update to that new size.
        // Do it without animation to more closely match the behavior in `UINavigationController`
        UIView.performWithoutAnimation {
            navigationBar.invalidateIntrinsicContentSize()
            navigationBar.layoutIfNeeded()
        }
        
        super.layoutSubviews()
        
        if let captionView = captionView as? NYTPhotoCaptionViewLayoutWidthHinting {
            captionView.preferredMaxLayoutWidth = bounds.width
        }
    }
}

